contrib/babelfishpg_tsql/sql/sys_functions.sql

```diff
@@ -3944,3 +3944,290 @@ RETURNS sys.SYSNAME AS $$
+    ELSIF date_arg_datatype IN ('sys.datetime2'::regtype, 'sys.datetimeoffset'::regtype) THEN
+        -- when datepart is {year, quarter, month} make use of AGE() function to find number of buckets
+        IF datepart IN ('year', 'quarter', 'month') THEN
+            date_difference_interval := AGE(date_trunc('day', date::timestamp), date_trunc('day', origin::timestamp));
+            years_diff := EXTRACT('Year' from date_difference_interval)::INT;
+            months_diff := EXTRACT('Month' from date_difference_interval)::INT;
+            CASE datepart
+            WHEN 'year' THEN
+                -- Here we are finding how many buckets we have to add in the origin so that we can reach to a bucket in whcih our date falls.
+                -- For cases where origin > date, we might end up in a bucket which exceeds date by 1 bucket. (Ex. date_bucket(year, 2, '2010-01-01', '2019-01-01')) 
+                required_bucket := years_diff/number;
+                result_date := origin::timestamp + concat(required_bucket * number, ' years')::interval;
+                IF result_date > date::timestamp THEN
+                    result_date = result_date - concat(number, ' years')::interval;
+                END IF;
+
+            WHEN 'month' THEN
+                required_bucket := (12 * years_diff + months_diff)/number;
+                result_date := origin::timestamp + concat(required_bucket * number, ' months')::interval;
+                IF result_date > date::timestamp THEN
+                    result_date = result_date - concat(number, ' months')::interval;
+                END IF;
+
+            WHEN 'quarter' THEN
+                years_diff := (12 * years_diff + months_diff)/3;
+                required_bucket := years_diff/number;
+                result_date := origin::timestamp + concat(required_bucket * number * 3, ' months')::interval;
+                IF result_date > date::timestamp THEN
+                    result_date = result_date - concat(number*3, ' months')::interval;
+                END IF;
+            END CASE;  
+        
+        -- when datepart is {week, day, hour, minute, second, millisecond} make use of built-in date_bin() postgresql function. 
+        ELSE
+            -- trunceting origin to millisecond before passing it to date_bin() function. 
+            -- store the difference between origin and trunceted origin to add it in the result of date_bin() function
+            date_difference_interval := concat(number, ' ', datepart)::INTERVAL;
+            millisec_trunc_diff_interval := (origin::timestamp - date_trunc('millisecond', origin::timestamp))::interval;
+            result_date = date_bin(date_difference_interval, date::timestamp, date_trunc('millisecond', origin::timestamp)) + millisec_trunc_diff_interval;
+
+            -- Filetering cases where the required bucket ends at date then date_bin() gives start point of this bucket as result. (Ex. query #4 of DATE_BUCKET_vu_prepare_v15) 
```

Comment:
Split this comment into multiple lines to make it more readable.

Reply:
Done